### PR TITLE
fix(package-rules): matchCurrentVersion with locked versions

### LINKED ISSUE #17670: Renovate suddenly started to mark dependencies as disabled
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

latest Docker image

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### If you're self-hosting Renovate, tell us what version of the platform you run.

15.2.2-ee

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Describe the bug

August 30th was the last time when MRs were created in our PHP app, using Renovate `32.179.1`. Next day, another scheduled pipeline used Renovate `32.185.1` and it closed MR created day before. After few days without any MRs, which seemed suspicious to me, I've run job with `LOG_LEVEL=debug` and saw that for `composer` manager there is a lot of `Dependency is disabled` logs. It seems like it's the reason why new MRs are not created.

This is weird because we did not change Renovate's config in terms of PHP packages. There was only small change in repositories' list handled by bot (change of namespace for frontend packages) that should not cause any side effects in PHP dependencies.

Below I'm adding example logs for `phpstan/phpstan` - current version is 1.8.4 so it should be updated, but was `disabled`. All other dependencies are disabled too.

### Relevant debug logs

<details><summary>Logs</summary>

```
...
Dependency is disabled (repository=getresponse/app, dependency=phpstan/phpstan)
...
               {
                 "depType": "require-dev",
                 "depName": "phpstan/phpstan",
                 "currentValue": "^1.7.2",
                 "datasource": "packagist",
                 "lockedVersion": "1.8.2",
                 "depIndex": 162,
                 "updates": [],
                 "skipReason": "disabled"
               },
...
```

</details>

### Have you created a minimal reproduction repository?

No reproduction repository

### PR DESCRIPTION
**Pull request by rarkins**
## Changes

Fixes `isUnconstrainedValue` to avoid `false` values when locked versions exist.

## Context

Closes #17670

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/util/package-rules/current-version.ts [hunk 1]
```diff
@@ -18,26 +18,29 @@ export class CurrentVersionMatcher extends Matcher {
     if (is.undefined(matchCurrentVersion)) {
       return null;
     }
-
-    if (is.nullOrUndefined(currentValue)) {
-      return false;
-    }
```
Comment (rarkins) [inline lib/util/package-rules/current-version.ts:24]: Before the class refactoring, we handled the case where `currentValue` was undefined but `lockedVersion` exists

File: lib/util/package-rules/current-version.ts [hunk 2]
```diff
@@ -18,26 +18,29 @@ export class CurrentVersionMatcher extends Matcher {
...
-
-    const isUnconstrainedValue = !!lockedVersion;
```
Comment (rarkins) [inline lib/util/package-rules/current-version.ts:26]: Post-refactoring, this was wrong and causing incorrect `false` results

File: lib/util/package-rules/current-version.ts [hunk 3]
```diff
@@ -18,26 +18,29 @@ export class CurrentVersionMatcher extends Matcher {
...
       matchCurrentVersionStr
     );
 
     if (matchCurrentVersionPred) {
-      return !(!isUnconstrainedValue && !matchCurrentVersionPred(currentValue));
+      const compareVersion = lockedVersion ?? currentVersion ?? currentValue;
+      return (
+        !is.nullOrUndefined(compareVersion) &&
+        matchCurrentVersionPred(compareVersion)
+      );
```
Comment (rarkins) [inline lib/util/package-rules/current-version.ts:34]: `matchCurrentVersion` should try to match against the version first, and only fall back to value if no version is found

**Comment by rarkins** [comment]
Earlier refactoring: #16865

**Review by secustor — APPROVED** [review]

**Review by viceice — APPROVED** [review]

**Pull request merged**